File: src/query_parallel_handler/query_parallel_handler.erl
-module(query_parallel_handler).

% elp:ignore W0012 W0040 E1599
-moduledoc {file, "../../docs/query_parallel_handler.md"}.

-export([
    start/0
]).

-define(PROGRESS_QUERY, ~"get_progress").

start() ->
    Cluster = cluster_1,
    {ok, #{workflow_execution := WE}} =
        temporal_sdk:start_workflow(Cluster, "default", query_parallel_handler_workflow),
    run_query(Cluster, WE, 0).

run_query(Cluster, WE, Count) when Count < 13 ->
    QR = temporal_sdk:query_workflow(Cluster, WE, ?PROGRESS_QUERY),
    inspect_query_response(QR, lists:concat(["progress after ", Count, " sec"])),
    timer:sleep(1_000),
    run_query(Cluster, WE, Count + 1);
run_query(Cluster, WE, _Count) ->
    temporal_sdk:await_workflow(Cluster, WE),
    QR = temporal_sdk:query_workflow(Cluster, WE, ?PROGRESS_QUERY),
    inspect_query_response(QR, "closed WF query").

inspect_query_response({ok, #{query_result := R}}, Header) ->
    io:fwrite("~s: ~p~n", [Header, R]);
inspect_query_response(Err, Header) ->
    io:fwrite("~s: ~p~n", [Header, Err]).

File: lib/query_parallel_handler/query_parallel_handler.ex
defmodule QueryParallelHandler do
  @external_resource "docs/query_parallel_handler.md"
  @moduledoc File.read!("docs/query_parallel_handler.md")

  @progress_query <<"get_progress">>

  def start() do
    cluster = :cluster_1

    {:ok, %{workflow_execution: we}} =
      TemporalSdk.start_workflow(cluster, "default", QueryParallelHandler.Workflow)

    run_query(cluster, we, 0)
  end

  defp run_query(cluster, we, count) when count < 13 do
    qr = TemporalSdk.Service.query_workflow(cluster, we, @progress_query)
    inspect_query_response(qr, "progress after #{count} sec")
    Process.sleep(1_000)
    run_query(cluster, we, count + 1)
  end

  defp run_query(cluster, we, _count) do
    TemporalSdk.await_workflow(cluster, we)
    qr = TemporalSdk.Service.query_workflow(cluster, we, @progress_query)
    inspect_query_response(qr, "closed WF query")
  end

  defp inspect_query_response({:ok, %{query_result: r}}, header),
    do: IO.puts("#{header}: #{inspect(r)}")

  defp inspect_query_response(err, header),
    do: IO.puts("#{header}: #{inspect(err)}")
end
